src/kokkos/plugin-SiPixelClusterizer/kokkos/gpuClustering.h

```diff
@@ -15,31 +15,32 @@
 namespace KOKKOS_NAMESPACE {
   namespace gpuClustering {
 
-#ifdef TODO
 #ifdef GPU_DEBUG
     __device__ uint32_t gMaxHit = 0;
 #endif
 
-    KOKKOS_INLINE_FUNCTION void countModules(uint16_t const* __restrict__ id,
-                                             uint32_t* __restrict__ moduleStart,
-                                             int32_t* __restrict__ clusterId,
-                                             int numElements) {
-      int first = blockDim.x * blockIdx.x + threadIdx.x;
-      for (int i = first; i < numElements; i += gridDim.x * blockDim.x) {
-        clusterId[i] = i;
-        if (InvId == id[i])
-          continue;
-        auto j = i - 1;
-        while (j >= 0 and id[j] == InvId)
-          --j;
-        if (j < 0 or id[j] != id[i]) {
-          // boundary...
-          auto loc = atomicInc(moduleStart, MaxNumModules);
-          moduleStart[loc + 1] = i;
-        }
+    KOKKOS_INLINE_FUNCTION void countModules(Kokkos::View<uint16_t const*, KokkosExecSpace> id,
+                                             Kokkos::View<uint32_t*, KokkosExecSpace> moduleStart,
+                                             Kokkos::View<int32_t*, KokkosExecSpace> clusterId,
+                                             int numElements,
+                                             const size_t index) {
+      clusterId[index] = index;
+      if (::gpuClustering::InvId == id[index])
+        return;
+      int j = index - 1;
+      while (j >= 0 and id[j] == ::gpuClustering::InvId)
+        --j;
+      if (j < 0 or id[j] != id[index]) {
+        // boundary... replacing atomicInc with explicit logic
+        auto loc = moduleStart[0];
+        if (moduleStart[0] >= ::gpuClustering::MaxNumModules)
+          moduleStart[0] = 0;
+        else
+          moduleStart[0] = moduleStart[0] + 1;
+        moduleStart[loc + 1] = index;
```

Comment:
I'm pretty sure the "reading of `loc`" and "updating `moduleStart[0]`" need to stay in one atomic transaction.

Reply:
Yes, you're right. I will change this. 

The Cuda `atomicInc` does
`((old >= val) ? 0 : (old+1))`
in one operation. This was used in the original. 

However, I don't see a similar Kokkos atomic. There is an `atomic_increment()` but it does not do the comparison. 
``` c
          auto loc = moduleStart[0];
          if( moduleStart[0] >= ::gpuClustering::MaxNumModules)
            Kokkos::atomic_assign(moduleStart,0);
          else:
            Kokkos::atomic_increment(moduleStart,0);
          moduleStart[loc + 1] = index;
```

Reply:
I think this will ensure  the proper atomics and behavior of the cuda version.
```c
auto loc = Kokkos::atomic_compare_exchange(moduleStart,::gpuClustering::MaxNumModules,0);
Kokkos::atomic_increment(moduleStart);
moduleStart[loc + 1] = index;
```

Reply:
Thanks, but I'm not sure if this is sufficient. E.g. if two threads hit into `atomic_compare_exchange` before either of them go to `atomic_increment`, they get the same value for `loc`, right?

The `atomic_fetch_add` would correctly combine the "return old value and increase", but misses the wrapping over `MaxNumModules`. On the other hand, I think the wrapping is mostly a sanity check to to prevent out-of-bounds access on `moduleStart`. And in fact, looking from the in-house CUDA-to-CPU translation header
https://github.com/cms-patatrack/cmssw/blob/a75d7c6abdf37434a737fc09aef05171b5e4d81a/HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h#L30-L36
it looks to me that the even the wrapping behavior is not needed, but preventing out-of-bounds is sufficient.

So maybe combining `atomic_fetch_add` with a local check `if (loc >= ::gpuClustering::MaxNumModules) loc = 0` (or `loc = ::gpuClustering::MaxNumModules-1`) could be sufficient?

@VinInn Could you confirm if I understood the intended use of `atomicInc()` correctly?

Reply:
OK, I wasn't quite sure if the wrapping check was just a precaution or if it was expected that wrapping is necessary.

Reply:
I think it was historical. it should be an assert actually. I personally do not like the behaviour of atomicInc...
I am not sure what your aim is here: targetting GPU, CPU serial or CPU MT? or trying CPU SIMD?

the logic is simple: if moduleId changes the module changes nd just increase the counter 
How? depends on the architecture and the concurrency model.

Reply:
Thanks @VinInn.

> I think it was historical. it should be an assert actually. I personally do not like the behaviour of atomicInc...

Ok, so the following would actually be the preferred behavior.
```cpp
auto loc = Kokkos::atomic_fetch_add(moduleStart, 1);
assert(loc < ::gpuClustering::MaxNumModules);
```

> I am not sure what your aim is here: targetting GPU, CPU serial or CPU MT? or trying CPU SIMD?

Portable code that would ideally support all of those. In practice for now (with Kokkos) that means supporting CPU serial and CUDA.